lib/hammer/mnesia.ex

```diff
@@ -0,0 +1,156 @@
+
+    quote do
+      @table unquote(table)
+
+      def child_spec(opts) do
+        %{
+          id: __MODULE__,
+          start: {__MODULE__, :start_link, [opts]},
+          type: :worker
+        }
+      end
+
+      def start_link(opts) do
+        opts = Keyword.put(opts, :table, @table)
+        Hammer.Mnesia.start_link(opts)
+      end
+
+      def hit(key, scale, limit, increment \\ 1) do
+        Hammer.Mnesia.hit(@table, key, scale, limit, increment)
+      end
+
+      def inc(key, scale, increment \\ 1) do
+        Hammer.Mnesia.inc(@table, key, scale, increment)
+      end
+
+      def set(key, scale, count) do
+        Hammer.Mnesia.set(@table, key, scale, count)
+      end
+
+      def get(key, scale) do
+        Hammer.Mnesia.get(@table, key, scale)
+      end
+    end
+  end
+
+  use GenServer
+
+  @doc """
+  Starts the process that creates and cleans the ETS table.
+
+  Accepts the following options:
+    - `:clean_period` for how often to perform garbage collection
+    - optional mnesia options from `:mnesia.create_table/2`
+    - optional `:debug`, `:spawn_opts`, and `:hibernate_after` `GenServer.options()`
+  """
+  def start_link(opts) do
+    {gen_opts, opts} = Keyword.split(opts, [:debug, :spawn_opt, :hibernate_after])
+    GenServer.start_link(__MODULE__, opts, gen_opts)
+  end
+
+  @doc false
+  def hit(table, key, scale, limit, increment) do
+    now = now()
+    full_key = full_key(key, scale, now)
+    count = :mnesia.dirty_update_counter(table, full_key, increment)
+
+    if count <= limit do
+      {:allow, count}
+    else
+      {_, _, expires_at} = full_key
+      retry_after = max(expires_at - now, 0)
+      {:deny, retry_after}
+    end
+  end
+
+  @doc false
+  def inc(table, key, scale, increment) do
+    :mnesia.dirty_update_counter(table, full_key(key, scale), increment)
+  end
+
+  @doc false
+  def set(table, key, scale, count) do
+    full_key = full_key(key, scale)
+    current_count = :mnesia.dirty_update_counter(table, full_key, 0)
+    :mnesia.dirty_update_counter(table, full_key, count - current_count)
+  end
+
+  @doc false
+  def get(table, key, scale) do
+    :mnesia.dirty_update_counter(table, full_key(key, scale), 0)
+  end
+
+  @compile inline: [full_key: 2]
+  defp full_key(key, scale) do
+    full_key(key, scale, now())
+  end
+
+  @compile inline: [full_key: 3]
+  defp full_key(key, scale, now) do
+    window = div(now, scale)
+    expires_at = (window + 1) * scale
+    {key, window, expires_at}
+  end
+
+  @compile inline: [now: 0]
+  defp now do
+    System.system_time(:millisecond)
+  end
+
+  @impl GenServer
+  def init(opts) do
+    {clean_period, opts} = Keyword.pop!(opts, :clean_period)
+    {table, mnesia_opts} = Keyword.pop!(opts, :table)
+
+    mnesia_opts = Keyword.merge(mnesia_opts, type: :set, attributes: [:key, :count])
+
+    case :mnesia.create_table(table, mnesia_opts) do
+      {:atomic, :ok} -> :ok
+      {:aborted, {:already_exists, _}} -> :ok
```

Comment:
Not sure what to do about this. Technically we should raise, but raising here would bring down the supervisor and then the whole application since it would hit 5 restarts per 3 seconds very quickly.

Reply:
I mean if you can t start it and is needed that seems like an issue?

Reply:
That's not what this is about. What you are asking about is probably the next line below that one, 129.

This is about a table that might exist but have a different structure.